dotnet/src/Experimental/Process.Abstractions/KernelProcessStepState.cs

```diff
@@ -31,18 +50,30 @@ public KernelProcessStepState(string name, string? id = null)
         this.Id = id;
         this.Name = name;
     }
+
+    /// <summary>
+    /// Registers a derived type for serialization. Types registered here are used by the KnownType attribute
+    /// to support DataContractSerialization of derived types as required to support Dapr.
+    /// </summary>
+    /// <param name="derivedType">A Type that derives from <typeref name="KernelProcessStepState"/></param>
+    public static void RegisterDerivedType(Type derivedType)
+    {
+        s_knownTypes.Add(derivedType);
+    }
 }
 
 /// <summary>
 /// Represents the state of an individual step in a process that includes a user-defined state object.
 /// </summary>
 /// <typeparam name="TState">The type of the user-defined state.</typeparam>
+[DataContract]
 public sealed record KernelProcessStepState<TState> : KernelProcessStepState where TState : class, new()
 {
     /// <summary>
     /// The user-defined state object associated with the Step.
     /// </summary>
-    public TState? State { get; set; }
+    [DataMember]
+    public TState? State { get; init; }
```

Comment:
Why not `[DataMember(Name = "state")]?  Truthfully, if there's not an external model to align with, I might not provide any `Name` override for any of the data-members...or if there is a particular reason to define an alternate contract name, wouldn't it apply here?